src/main/java/org/apache/mesos/scheduler/recovery/SimpleRecoveryPlanManager.java

```diff
@@ -0,0 +1,325 @@
+package org.apache.mesos.scheduler.recovery;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.TextFormat;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.mesos.Protos;
+import org.apache.mesos.offer.DefaultOfferRequirementProvider;
+import org.apache.mesos.offer.InvalidRequirementException;
+import org.apache.mesos.offer.OfferRequirementProvider;
+import org.apache.mesos.scheduler.plan.*;
+import org.apache.mesos.scheduler.recovery.constrain.LaunchConstrainer;
+import org.apache.mesos.scheduler.recovery.monitor.FailureMonitor;
+import org.apache.mesos.specification.DefaultTaskSpecification;
+import org.apache.mesos.specification.InvalidTaskSpecificationException;
+import org.apache.mesos.state.StateStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+/**
+ * {@link SimpleRecoveryPlanManager} enables monitoring and management of recovery plan.
+ *
+ * This is an implementation of {@code PlanManager} that performs task recovery using dynamically generated
+ * {@code Plan}. {@link SimpleRecoveryPlanManager} tracks currently failed (permanent) and stopped (transient) tasks,
+ * generates a new {@link DefaultRecoveryBlock} for them and adds them to the recovery Plan, if not already added.
+ */
+public class SimpleRecoveryPlanManager implements PlanManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SimpleRecoveryPlanManager.class);
+
+    protected volatile Plan plan;
+    protected volatile PhaseStrategy phaseStrategy;
+
+    private final StateStore stateStore;
+    private final TaskFailureListener failureListener;
+    private final RecoveryRequirementProvider offerReqProvider;
+    private final FailureMonitor failureMonitor;
+    private final LaunchConstrainer launchConstrainer;
+    private final UUID phaseId = UUID.randomUUID();
+
+    public SimpleRecoveryPlanManager(
+            StateStore stateStore,
+            TaskFailureListener failureListener,
+            RecoveryRequirementProvider offerReqProvider,
+            LaunchConstrainer launchConstrainer,
+            FailureMonitor failureMonitor) {
+        this.stateStore = stateStore;
+        this.offerReqProvider = offerReqProvider;
+        this.failureMonitor = failureMonitor;
+        this.launchConstrainer = launchConstrainer;
+        this.failureListener = failureListener;
+    }
+
+    @VisibleForTesting
+    protected synchronized void refreshPlan(List<Block> dirtiedAssets) {
```

Comment:
It is not. This is called from `getCurrentBlock()` method which is called from `PlanCoordinator` (intended). But because `getCurrentBlock()` is a public method and can be called from anywhere else, I did some defensive programming to ensure that this is thread safe.